text/0000-cargo-workspace.md

```diff
@@ -0,0 +1,345 @@
+get out of sync if it were allowed. By ensuring that all crates have edges to
+each other in a workspace Cargo can prevent this situation and guarantee robust
+builds no matter where they're executed in the workspace.
+
+To alleviate misconfiguration Cargo will emit an error if the two properties
+above hold for any crate attempting to be part of a workspace. For example, if
+the `package.workspace` key is specified, but the crate is not a workspace root
+or doesn't point back to the original crate an error is emitted.
+
+### Implicit relations
+
+The combination of the `package.workspace` key and `[workspace]` table is enough
+to specify any workspace in Cargo. Having to annotate all crates with a
+`package.workspace` parent or a `workspace.members` list can get quite tedious,
+however! To alleviate this configuration burden Cargo will allow these keys to
+be implicitly defined in some situations.
+
+The `package.workspace` can be omitted if it would only contain `../` (or some
+repetition of it). That is, if the root of a workspace is hierarchically the
+first `Cargo.toml` with `[workspace]` above a crate in the filesystem, then that
+crate can omit the `package.workspace` key.
+
+Next, a crate which specifies `[workspace]` **without a `members` key** will
+transitively crawl `path` dependencies to fill in this key. This way all `path`
+dependencies (and recursively their own `path` dependencies) will inherently
+become the default value for `workspace.members`.
+
+Note that these implicit relations will be subject to the same validations
+mentioned above for all of the explicit configuration as well.
+
+### Workspaces in practice
+
+Many Rust projects today already have `Cargo.toml` at the root of a repository,
+and with the small addition of `[workspace]` in the root a workspace will be
+ready for all crates in that repository. For example:
+
+* An FFI crate with a sub-crate for FFI bindings
+
+  ```
+  Cargo.toml
+  src/
+  foo-sys/
+    Cargo.toml
+    src/
+  ```
+
+* A crate with multiple in-tree dependencies
+
+  ```
+  Cargo.toml
+  src/
+  dep1/
+    Cargo.toml
+    src/
+  dep2/
+    Cargo.toml
+    src/
+  ```
+
+Some examples of layouts that will require extra configuration, along with the
+configuration necessary, are:
+
+* Trees without any root crate
+
+  ```
+  crate1/
+    Cargo.toml
+    src/
+  crate2/
+    Cargo.toml
+    src/
+  crate3/
+    Cargo.toml
+    src/
+  ```
+
+  these crates can all join the same workspace via a `Cargo.toml` file at the
+  root looking like:
+
+  ```toml
+  [workspace]
+  members = ["crate1", "crate2", "crate3"]
+  ```
+
+* Trees with multiple workspaces
+
+  ```
+  ws1/
+    crate1/
+      Cargo.toml
+      src/
+    crate2/
+      Cargo.toml
+      src/
+  ws2/
+    Cargo.toml
+    src/
+    crate3/
+      Cargo.toml
+      src/
+  ```
+
+  The two workspaces here can be configured by placing the following in the
+  manifests:
```

Comment:
I'm not sure I understand what change you'd like as a result of these comments then? The intention is that of course two independent trees can have workspaces that don't mess with one another, and the point of this example is simply to show that it can happen in one repo.